pkg/registry/common/connect/ns_server.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connect
+
+import (
+	"context"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/common/clienturl"
+	"github.com/networkservicemesh/sdk/pkg/registry/core/adapters"
+)
+
+type nsCacheEntry struct {
+	expirationTimer *time.Timer
+	client          registry.NetworkServiceRegistryClient
+}
+
+type connectNSServer struct {
+	dialOptions       []grpc.DialOption
+	clientFactory     func(ctx context.Context, cc grpc.ClientConnInterface) registry.NetworkServiceRegistryClient
+	cache             nsClientMap
+	connectExpiration time.Duration
```

Comment:
Why are we setting a single arbitrary global timer here rather than taking the timing information from the events coming in.  When we proxy registration, that registration has an expiration timer, when we do a find, the responses have expiration timers... why introduce arbitrary new global ones into the system?

Reply:
> Why introduce arbitrary new global ones into the system?

Because if we got for example from domain1 NSE with expiration_time = 1 year, then we'll never close connection...

Reply:
@edwarnicke In short

### Problems of way to use expiration time of resource

## 1. Resources with long expiration time.

As I said above few times, if connect got NSE/NS with expiration_time > 1 year and so on then we'll never close the connection. The worst thing of this that the connection may be not even used...


## 2. The complex system of monitoring

By this point I mean the cases when we have two or more resources from one domain with different expiration time... Image the case:
User did these steps:
```
1. Register NSE `nse-2` with expiration_time 1 minute
2. Register NSE `nse-3` with expiration_time 20 minutes 
3. Register NSE `nse-1` with expiration_time 1 day // this will be selected as expiration time of connection
4. Unregister NSE `nse-1` 
```
In this case, we'll need to select the next expiration time  (20 minutes) and for this, we'll need to integrate the whole list of records... And so on cases.

## 3. Leak of connections

Let's imagine the next case: 

1. User requests `ns-1@domain1` in time t1.
2. DNS result of domain1 maybe url1 
3. Registry `connect` add to cache url1 and sets expiration time = ns-1.expiration_time(for example 1day)
4. User requests `ns-2@domain1`  in time t2.
5. DNS result of domain1 in this time maybe url2
6. Registry `connect` add to cache url2 and sets expiration time = ns-2.expiration_time(for example 1day) //!!! We have two connections when one of them may be not using!

**Actual**: We have 2 connections when one connection is not used more.
**Expected**: Do not need unused connections.

###  Solution

Use the map of connection where expiration time depends on user interested. (User doing many requests to domain1 OK store connection to domain1 in the cache. User did one time connect to domain2 and do not use this connection OK lets just remove it and so on)

@edwarnicke Please let me know If I missing something.

Reply:
Good point on long expire times on resources.  I like your proposed solution in principle (somehow detecting user interest to determine how long to hang onto a connection), what in concrete terms did you have in mind?  How do we detect user interest in a thing?

Reply:
Actually... wasn't part of the point of monitor to 'monitor' registered resources so we could quickly reregister if they disappear?

Reply:
Can we just use a current approach that if a connection is loaded from a cache then just update delete timer 
 https://github.com/networkservicemesh/sdk/pull/374/files#diff-c4e8f787c3f17d8f27d19a98904cf44bR76? :)

Reply:
Hmm... could we move the connExpiration timer to a constant rather than a parameter the user has to think about?

Reply:
In this case, we'll need to return to options because this parameter should be modifiable for testing goals.

Reply:
Options are fine :)

Reply:
Done :)